Improved NumPy ndarray <--> BlockMatrix conversion

jbloom22 opened the pull request:
An overhaul of our NumPy ndarray and BlockMatrix conversion, proceeding through binary on disk compatibly with the NumPy `tofile` and `fromfile` functions. For expediency, I still use Breeze matrix as an (internal) intermediate, resulting in an extra copy and max 2^31 size. Longer term, we should avoid routing through Breeze, add compression, etc, but this gets us a lot of bang for the buck in the face of intolerable py4j slowness. In fact, on laptop, the local byte transport piece is about 50x faster through disk than the "faster" direction of py4j (java to python via byte array), so now 90% of the time is spent localizing and distributing serialized blocks. E.g. reading a 8192 x 8192 block matrix with 4 blocks and converting to NumPy takes about 12s. Converting the local matrix to NumPy takes a bit over 1s. The current "slower" py4j direction (python to java) falls over even on tiny matrices.

Main changes:
- added `tofile, to_numpy, fromfile, from_numpy` to BlockMatrix, which are explained in the docs with examples and tested in `test_linalg`.
- deleted all breeze related functions in utils.java
- added a `StreamRawBlockBufferSpec` (better name?) which behaves like the `StreamBlockBufferSpec` except that it only writes the (raw) blocks without adding length data to the stream. This allows for re-using readDoubles and writeDoubles as implemented in BlockingBuffer.
- added readDoubles and writeDoubles using this buffer spec on RichDenseMatrixDouble, and a test.
- along the way, added a `hl.tmp_dir()` function to allow users to inspect the tmp_dir used with `BlockMatrix.from_entry_expr`, as noted in that documentation.

jbloom22 commented:
Ack, this isn't working as written in cluster mode on GCP due to different meaning of file names locally and through Hadoop. I'll think on it but would also appreciate feedback before I go further on this branch.

tpoterba commented:
take a look at Jackie's fasta branch. You'll need to use URIs to get the path Python can read.

jbloom22 commented:
I've removed the Python `tempfile` approach in favor of adding `new_local_temp_file` to utils and a corresponding function to HailContext in Scala, which currently hardcodes `file:///temp` as the local temp directory. It may be more natural to have a localTmpDir on HailContext like we have tmpDir. 
I see there is a notion of local temp files on TempDir on the Scala side, but it doesn't seem to be used on the Python side. I also don't see if/where we wipe temp files on exit.

In any case, I've tested that now it all works nicely on GCP, so ready for feedback/review.

I think factoring through `tofile` and `fromfile` is useful for wider interoperability for the same reason that NumPy exposes them, but it's also good if you don’t want to actually load the NumPy array into driver memory but just save it to read/copy later, or to load it multiple time without recomputing the BlockMatrix. And I've provided the simpler interface of `to_numpy` and `from_numpy` for the common case.

I suspect that (de)serializing over the network and building the local matrix dominates local read/write, so that using a socket isn't going to do much better. I can profile more closely if/when we feel it's high priority to make this faster still.

tpoterba commented:
TempDir has local files, that's totally the thing to use

jbloom22 commented:
Yes but then TempDir should be on HailContext rather than just a string. I’ll look into making that change.

jbloom22 commented:
I'm now using mkdtemp and atexit, ready for review.

jbloom22 commented:
I've verified that the temp directory is blown away at exit, and re-rested everything on GCP.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -77,47 +154,44 @@ def from_numpy(cls, ndarray, block_size=None):
-        ndarray = ndarray.astype(np.float64)
         if ndarray.ndim != 2:
             raise FatalError("from_numpy: ndarray must have two axes, found shape {}".format(ndarray.shape))
         n_rows, n_cols = ndarray.shape[0:2]
```
You don't need [0:2], you just checked it had ndim == 2.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -51,23 +51,100 @@ def read(cls, path):
+
+        where the local temp file is created on the driver using Python's
+        `tempfile <https://docs.python.org/3/library/tempfile.html>`__
+        library and deleted at exit.
```
So, loosely I think of documented things as part of the interface.  We should have the option to change the implementation, and what's more, this is probably more information than the users need.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -51,23 +51,100 @@ def read(cls, path):
+        The number of entries must be less than :math:`2^{31}`.
+
+        Use :meth:`BlockMatrix.write` and :meth:`BlockMatrix.read` to save and
+        restore block matrices.
```
save and load.  Also, you might say in the case that you don't need to interoperate with numpy, in particular, because read/write read and write in parallel.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -77,47 +154,44 @@ def from_numpy(cls, ndarray, block_size=None):
         -----
         For shuffle resiliency, this functions writes the resulting block matrix
         to disk and then reads the result. By specifying a path, this block
         matrix can be read again using :meth:`BlockMatrix.read`. Otherwise, a
```
Again, too much detail, although perhaps we can promise it is shuffle free (although I could imagine we'd change that if we had a reliable shuffle strategy).

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -77,47 +154,44 @@ def from_numpy(cls, ndarray, block_size=None):
-        temporary file is used and then deleted when :func:`hail.stop`
-        is called or the program terminates.
+        file is created in the temp directory given by :func:`hail.tmp_dir` and
+        then deleted when :func:`hail.stop` is called or the program terminates.
```
How does it get deleted when stop is called?

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -288,23 +378,90 @@ def filter(self, rows_to_keep, cols_to_keep):
+
+        Notes
+        -----
+        The number of entries must be less than :math:`2^{31}`.
```
This comment is unnecessary, it already is.

jbloom22 replied:
? Both a BlockMatrix and a NumPy array (in 64bit Python) can have more than 2^31 entries.

cseed replied:
Ah, because you go through breeze.  Nevermind.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -288,23 +378,90 @@ def filter(self, rows_to_keep, cols_to_keep):
+
+        Use :meth:`.utils.hadoop_copy` to copy files between local and non-local
+        storage. Use :meth:`BlockMatrix.write` and :meth:`BlockMatrix.read` to
+        save and restore block matrices.
```
Again, why?

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -288,23 +378,90 @@ def filter(self, rows_to_keep, cols_to_keep):
         Notes
         -----
-        The resulting ndarray will have shape ``(n_rows, n_cols)``.
+        The number of entries must be less than :math:`2^{31}`.
```
It already is.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -288,23 +378,90 @@ def filter(self, rows_to_keep, cols_to_keep):
+
+        where the local temp file is created on the driver using Python's
+        `tempfile <https://docs.python.org/3/library/tempfile.html>`__
+        library and deleted at exit.
```
Same comment.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -77,47 +154,44 @@ def from_numpy(cls, ndarray, block_size=None):
         matrix can be read again using :meth:`BlockMatrix.read`. Otherwise, a
-        temporary file is used and then deleted when :func:`hail.stop`
-        is called or the program terminates.
+        file is created in the temp directory given by :func:`hail.tmp_dir` and
```
You don't call `tmp_dir`.

cseed commented on python/hail/context.py:
```diff
@@ -207,3 +207,8 @@ def get_reference(name):
             ReferenceGenome._from_java(Env.hail().variant.ReferenceGenome.getReference(name))
         )
+
+def tmp_dir():
```
Unused.

cseed commented on src/main/scala/is/hail/io/RowStore.scala:
```diff
@@ -51,6 +51,12 @@ final class StreamBlockBufferSpec extends BlockBufferSpec {
   def buildOutputBuffer(out: OutputStream): OutputBlockBuffer = new StreamBlockOutputBuffer(out)
 }
 
+final class StreamRawBlockBufferSpec(blockSize: Int) extends BlockBufferSpec {
```
I won't ask you to change this but it makes me uneasy.

jbloom22 commented:
bump

cseed commented on src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala:
```diff
@@ -21,30 +37,43 @@ object RichDenseMatrixDouble {
+    hc.hadoopConf.readFile(path) { is =>
+      val in = rawBufferSpec.buildInputBuffer(is)
+
+      in.readDoubles(data)
```
You need to check the return value here.

jbloom22 replied:
I'm confused. There is no return value. readDoubles will only return if `data.length` doubles have been read.

cseed commented on src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala:
```diff
@@ -21,30 +37,43 @@ object RichDenseMatrixDouble {
     val data = new Array[Double](rows * cols)
     in.readDoubles(data)
```
You need to check the return value here.

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -51,23 +51,94 @@ def read(cls, path):
+        reads a binary file of float64 values in row-major order, such as that
+        produced by `numpy.tofile
+        <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.fromfile.html>`__
+        or :meth:`BlockMatrix.tofile`.
```
That link is broken.

What's more, from the fromfile documentation:

 > Do not rely on the combination of tofile and fromfile for data storage, as the binary files generated are are not platform independent. In particular, no byte-order or data-type information is saved.

You should make it the same point these are not portable and should only be used for interoperating with numpy, not storage.

You should also make it clear the matrix written on the numpy side had to have type float64 and that this is not checked (nor is it possible to check given the file format).

cseed commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -77,47 +155,41 @@ def from_numpy(cls, ndarray, block_size=None):
-        temporary file is used and then deleted when :func:`hail.stop`
-        is called or the program terminates.
+        By specifying a path, the resulting block matrix can be re-loaded with
+        :meth:`BlockMatrix.read`.
```
OK, so I don't really like this because it forces us to write, and we might have another strategy to do the conversion in the future.  How important is it that it gets documented?

The user always has the option of writing explicitly if that's what they want.

cseed replied:
Also, does it overwrite when you specify the path?

jbloom22 commented:
- separated `BlockMatrix.from_entry_expr` (a classmethod) and `BlockMatrix.write_from_entry_expr`  (a staticmethod)
- added buffers specifically for handling Doubles.
- moved RichDenseMatrixDouble to its own suite and test the condition of the file being shorten than expected

cseed commented on src/main/scala/is/hail/utils/richUtils/RichInputStream.scala:
```diff
@@ -9,7 +10,9 @@ class RichInputStream(val in: InputStream) extends AnyVal {
     while (read < n) {
       val r = in.read(to, toOff + read, n - read)
-      assert(r > 0)
+      assert( r != 0)
```
Spacing.

cseed commented on src/main/scala/is/hail/utils/richUtils/RichInputStream.scala:
```diff
@@ -9,7 +10,9 @@ class RichInputStream(val in: InputStream) extends AnyVal {
-      assert(r > 0)
+      assert( r != 0)
+      if (r == -1)
+        fatal("Reached end of file prematurely")
```
"Premature end of file"

cseed commented:
So I can't resist asking, what's missing for LMMs now?

jbloom22 commented:
I'm currently expanding BlockMatrix binary ops to work naturally between BlockMatrix and 2d ndarray, with the useful ones for LMM being broadcasting ndarray over rows or columns of BlockMatrix. That'll round out the linear algebra, so that next week I'll begin refactoring the pipeline, with the interesting bits being the form of Python form of the small "global model" (initially using Scala side as black box) and then applying the local model to the pair or RowMatrices `X` and `PX` (read from BlockMatrices) to get per variant results.

cseed commented:
You mean 1darrays?

jbloom22 commented:
I want BlockMatrices to behave like 2d ndarrays. So I promote a 1d ndarray with shape (n) to a 2d array with shape (1, n) to get same broadcasting behavior as would occur in NumPy, e.g. added to each row. To add to each column, the 1d ndarray should be reshaped to 2d ndarray with shape (n, 1) first.
